engine/src/logger/instance.cpp

```diff
@@ -0,0 +1,288 @@
+#include <logger/instance.hpp>
+#include <algorithm>
+#include <condition_variable>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+#if defined(_WIN32)
+	#include <Windows.h>
+#endif
+
+namespace gen::logger {
+namespace {
+namespace fs = std::filesystem;
+
+void append_timestamp(std::string& out, Clock::time_point const& timestamp, Timestamp const mode) {
+	static auto s_mutex{std::mutex{}};
+	static constexpr std::size_t buf_size_v{64};
+	auto buffer = std::array<char, buf_size_v>{};
+	auto const time = Clock::to_time_t(timestamp);
+	// std::localtime / std::gmtime are not thread safe. cppref:
+	//  pointer to a static internal std::tm object on success, or null pointer otherwise.
+	// The structure may be shared between std::gmtime, std::localtime, and std::ctime,
+	// and may be overwritten on each invocation.
+	auto lock = std::unique_lock{s_mutex};
+	// NOLINTNEXTLINE
+	auto const tm_struct = mode == Timestamp::eUtc ? *std::gmtime(&time) : *std::localtime(&time);
+	lock.unlock();
+	std::strftime(buffer.data(), buffer.size(), "%F %T", &tm_struct);
+	out.append(buffer.data());
+}
+
+void append_location(std::string& out, std::source_location const& location, Location const mode) {
+	auto const path = [&] {
+		auto ret = fs::path{location.file_name()};
+		if (mode == Location::eFilename) { ret = ret.filename(); }
+		return ret.generic_string();
+	}();
+	out.append(path);
+}
+
+struct Formatter {
+	static constexpr auto open_v{'{'};
+	static constexpr auto close_v{'}'};
+
+	// snapshot of formatting-related Config (at the time of Formatter construction).
+	// we make copies instead of referring to Config to avoid mutex locks,
+	// since users can change the active config at any time.
+	struct Data {
+		// default format string size is 57, which will cause std::string to heap allocate.
+		// thus we use a fixed capacity stack string instead.
```

Comment:
Is there any means to avoid this limitation or are we fully locked in by std::string?

Reply:
Can't think of any means to avoid it except using a third party library for strings. We will need lots of fixed strings throughout UI code anyway, IMO having that class and building upon it later as needed is a better option for this app.

Reply:
[Context](https://github.com/ocornut/imgui/blob/master/docs/FAQ.md#q-how-can-i-interact-with-standard-c-types-such-as-stdstring-and-stdvector) for "lots of fixed strings in UI code":

> Dear ImGui applications often need to make intensive use of strings. It is expected that many of the strings you will pass to the API are raw literals (free in C/C++) or allocated in a manner that won't incur a large cost on your application. Please bear in mind that using std::string on applications with a large amount of UI may incur unsatisfactory performances. Modern implementations of std::string often include small-string optimization (which is often a local buffer) but those are not configurable and not the same across implementations.

Reply:
I agree, a fixed string class will be super useful down the line and I do think we should keep it; my biggest concern is that a 57-character limit is quite limiting for a log message as shown below in the example. How hard would it be to implement a third-party string solution? If it would take longer than 1-2 hours I'd say we should say whatever ever and deal with it but I'd like to hear your thoughts if you think it'd be feasible to bring something in that can circumvent the formatting cutoff.

Also for reference, this is 55 characters counting spaces.
"I hope you are enjoying your time finding this helpful."

Reply:
The limit is for the format specification, not the log string, that is unlimited.

Reply:
```cpp
static constexpr std::string_view default_format_v{"[{level}][T{thread}] {message} [{timestamp}] [{location}]"};
// ...
FixedString<format_size_v> format{default_format_v};
```
This part ^.

Reply:
Aaaaaah, ok, that clears things up much more. Sorry for my misunderstanding! I have no issue with the code then.